src/Microsoft.DotNet.Wpf/src/PresentationFramework/Microsoft/Win32/CommonItemDialog.cs

```diff
@@ -0,0 +1,693 @@
+
+        private protected virtual void PrepareDialog(IFileDialog dialog)
+        {
+            if (!string.IsNullOrEmpty(InitialDirectory))
+            {
+                IShellItem initialDirectory = ShellUtil.GetShellItemForPath(InitialDirectory);
+                if (initialDirectory != null)
+                {
+                    // Setting both of these so the dialog doesn't display errors when a remembered folder is missing.
+                    dialog.SetDefaultFolder(initialDirectory);
+                    dialog.SetFolder(initialDirectory);
+                }
+            }
+
+            dialog.SetTitle(Title);
+            dialog.SetFileName(CriticalFileName);
+
+            // Only accept physically backed locations.
+            FOS options = _dialogOptions.Value | FOS.FORCEFILESYSTEM;
+            dialog.SetOptions(options);
+
+            IList<FileDialogCustomPlace> places = CustomPlaces;
+            if (places != null && places.Count != 0)
+            {
+                foreach (FileDialogCustomPlace customPlace in places)
+                {
+                    IShellItem shellItem = ResolveCustomPlace(customPlace);
+                    if (shellItem != null)
+                    {
+                        try
+                        {
+                            dialog.AddPlace(shellItem, FDAP.BOTTOM);
+                        }
+                        catch (ArgumentException)
+                        {
+                            // The dialog doesn't allow some ShellItems to be set as Places (like device ports).
+                            // Silently swallow errors here.
+                        }
+                    }
+                }
+            }
+        }
+
+        // The FileOk event expects all properties to be set, but if the event is cancelled, they need to be reverted.
+        // This method is called inside a try block, and inheritors can store any data to be reverted in the revertState.
+        private protected virtual bool TryHandleItemOk(IFileDialog dialog, out object revertState)
+        {
+            revertState = null;
+            return true;
+        }
+
+        // This method is called inside a finally block when OK event was cancelled.
+        // Inheritors should revert properties to the state before the dialog was shown, so that it can be shown again.
+        private protected virtual void RevertItemOk(object state) { }
+
+        #endregion
+
+        //---------------------------------------------------
+        //
+        // Internal Properties
+        //
+        //---------------------------------------------------
+        #region Internal Properties
+
+        //   If multiple files are selected, we only return the first filename.
+        /// <summary>
+        ///  Gets a string containing the full path of the file selected in 
+        ///  the file dialog box.
+        /// </summary>
+        private protected string CriticalFileName
```

Comment:
We should change these from File -> Item for both properties and members.
Like MutableFileNames, CriticalFileName, _fileNames, etc.

Reply:
But these are strings, not `IShellItem`s

Reply:
IMHO, this will be consistent with the approach we are taking. Like 'CriticalItemName' makes more sense in CommonItemDialog and is generic enough for both files and folders.

Reply:
Updated in https://github.com/dotnet/wpf/pull/7916/commits/ef7eb70b508066dffc48c04f7683b5d437804d43